Move ReplicaSet controller to use apps/v1beta2 APIs

liyinan926 opened the pull request:
**What this PR does / why we need it**:
This PR moved the `ReplicaSet` controller to use `apps/v1beta2` APIs.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: #51049.

**Special notes for your reviewer**:

**Release note**:
```release-note
 Moved the ReplicaSet controller to use apps/v1beta2 APIs
```

liyinan926 commented:
/unassign @mbohlool 
/assign @lavalamp

janetkuo commented:
/ok-to-test

janetkuo commented on pkg/apis/apps/install/install.go:
```diff
@@ -37,7 +37,7 @@ func Install(groupFactoryRegistry announced.APIGroupFactoryRegistry, registry *r
 		&announced.GroupMetaFactoryArgs{
 			GroupName:                  apps.GroupName,
-			VersionPreferenceOrder:     []string{v1beta1.SchemeGroupVersion.Version, v1beta2.SchemeGroupVersion.Version},
+			VersionPreferenceOrder:     []string{v1beta2.SchemeGroupVersion.Version, v1beta1.SchemeGroupVersion.Version},
```
Does this change the default storage version? Will it work against <1.8 kubectl?
cc @kubernetes/sig-api-machinery-misc

dixudx replied:
@janetkuo This is just the preferences order. It will not break anything < 1.8 kubectl.

liggitt replied:
we cannot change the storage version to v1beta2 until 1.9 (TestEtcdStoragePath integration test will fail if this affected the storage version)

liyinan926 replied:
OK, reverted this change.

dixudx commented on cmd/kube-controller-manager/app/apps.go:
```diff
@@ -22,9 +22,23 @@ package app
 )
 
+func startReplicaSetController(ctx ControllerContext) (bool, error) {
+	if !ctx.AvailableResources[schema.GroupVersionResource{Group: "extensions", Version: "v1beta1", Resource: "replicasets"}] {
```
`extensions/v1beta1 replicasets`? Aren't we moving to a new version? We've already bumped to `apps/v1beta2`.

liyinan926 replied:
Yep. Missed this. Fixed.

dixudx commented on pkg/controller/replicaset/replica_set.go:
```diff
@@ -26,21 +26,20 @@ import (
 	"time"
 
 	"github.com/golang/glog"
+	apps "k8s.io/api/apps/v1beta2"
```
better to use `appsv1beta2` for consistency and management.

dixudx commented on cmd/kube-controller-manager/app/apps.go:
```diff
@@ -22,9 +22,23 @@ package app
+		return false, nil
+	}
+	go replicaset.NewReplicaSetController(
+		ctx.InformerFactory.Apps().V1beta2().ReplicaSets(),
```
This will cause error due to above declaration `"extensions/v1beta1"`. Change above.

liyinan926 replied:
Fixed.

dixudx commented:
@liyinan926 Please also bump new version in test/e2e/apps/replica_set.go. And update those generated files later.

dixudx commented on test/e2e/apps/replica_set.go:
```diff
@@ -35,7 +36,33 @@ import (
+	}
+}
+
+func newExtensionsV1beta1RS(rsName string, replicas int32, rsPodLabels map[string]string, imageName string, image string) *extensions.ReplicaSet {
```
Better to add a `TODO` to remove this method and use `newRS` instead when corecontroller `Deployment` get bumped to apps/v1beta2?

liyinan926 replied:
Done.

crimsonfaith91 commented:
/retest

liyinan926 commented:
/test pull-kubernetes-bazel

liyinan926 commented:
/test pull-kubernetes-unit

liyinan926 commented:
Finally got all Jenkins job to pass. @janetkuo @dixudx PTAL. Thanks!

dixudx commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:
```diff
@@ -210,6 +210,8 @@ func init() {
 		Rules: []rbac.PolicyRule{
 			rbac.NewRule("get", "list", "watch", "update").Groups(extensionsGroup).Resources("replicasets").RuleOrDie(),
 			rbac.NewRule("update").Groups(extensionsGroup).Resources("replicasets/status").RuleOrDie(),
+			rbac.NewRule("get", "list", "watch", "update").Groups(appsGroup).Resources("replicasets").RuleOrDie(),
```
No need to add 2 new Rules, just amend L211-212 to use `.Groups(extensionsGroup ,appsGroup)` instead, since the rules are exactly the same.

liggitt replied:
any of the rules referencing replicasets should have the appsGroup added

liyinan926 replied:
Done.

dixudx reviewed (commented):
@liyinan926 We also need to bump new version `apps/v1beta2` to below test yamls,

1. [hack/testdata/redis-slave-replicaset.yaml](https://github.com/kubernetes/kubernetes/blob/master/hack/testdata/redis-slave-replicaset.yaml#L1);
2. [hack/testdata/frontend-replicaset.yaml](https://github.com/kubernetes/kubernetes/blob/master/hack/testdata/frontend-replicaset.yaml#L1);

Correspondingly related codes using `extensions.replicaSet` in `pkg/kubectl` has to get bumped to `apps/v1beta2`.

@janetkuo Correct me if I'm wrong.

liyinan926 commented:
@dixudx @liggitt @janetkuo Is it better to have changes to `pkg/kubectl` in a separate PR, or it makes more sense to put them into this PR?

dixudx commented:
/test pull-kubernetes-unit

dixudx commented:
> Personally I feel it's cleaner and more manageable to have a separate PR for kubectl changes, which can be done independently from controller changes as both are just clients to CRUD ReplicaSet objects.

@liyinan926 SGTM. I'd give @liggitt and @janetkuo a few hours to chime in.

liyinan926 commented:
@janetkuo @liggitt PR updated. I intentionally left changes to `pkg/kubectl` out in this PR. As discussed above, IMO those changes are better to be done independently in a separate PR. Thoughts? 
PTAL. Thanks!

liggitt commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -159,7 +161,7 @@ func newPod(name string, rs *extensions.ReplicaSet, status v1.PodPhase, lastTran
 	if properlyOwned {
 		var trueVar = true
-		controllerReference = metav1.OwnerReference{UID: rs.UID, APIVersion: "v1beta1", Kind: "ReplicaSet", Name: rs.Name, Controller: &trueVar}
+		controllerReference = metav1.OwnerReference{UID: rs.UID, APIVersion: "v1beta2", Kind: "ReplicaSet", Name: rs.Name, Controller: &trueVar}
```
should be `apps/v1beta2`, right?

liyinan926 replied:
Done.

liggitt commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -173,10 +175,10 @@ func newPod(name string, rs *extensions.ReplicaSet, status v1.PodPhase, lastTran
 	pods := []v1.Pod{}
 	var trueVar = true
-	controllerReference := metav1.OwnerReference{UID: rs.UID, APIVersion: "v1beta1", Kind: "ReplicaSet", Name: rs.Name, Controller: &trueVar}
+	controllerReference := metav1.OwnerReference{UID: rs.UID, APIVersion: "v1beta2", Kind: "ReplicaSet", Name: rs.Name, Controller: &trueVar}
```
missing `apps` group

liyinan926 replied:
Done.

liggitt commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -357,15 +359,15 @@ func TestStatusUpdatesWithoutReplicasChange(t *testing.T) {
 
 	// This response body is just so we don't err out decoding the http response, all
 	// we care about is the request body sent below.
-	response := runtime.EncodeOrDie(testapi.Extensions.Codec(), &extensions.ReplicaSet{})
```
why did this change? I'd expect testapi.Apps.Codec()

liyinan926 replied:
`testapi.Apps` defaults to `apps/v1beta1`. This can only be changed if we change https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/apps/install/install.go#L40. But doing so changes the default storage version.

liggitt commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -357,15 +359,15 @@ func TestStatusUpdatesWithoutReplicasChange(t *testing.T) {
-	updatedRc := runtime.EncodeOrDie(testapi.Extensions.Codec(), rs)
-	fakeHandler.ValidateRequest(t, testapi.Extensions.ResourcePath(replicaSetResourceName(), rs.Namespace, rs.Name)+"/status", "PUT", &updatedRc)
+	updatedRc := runtime.EncodeOrDie(codec, rs)
+	fakeHandler.ValidateRequest(t, resourcePath(replicaSetResourceName(), rs.Namespace, rs.Name)+"/status", "PUT", &updatedRc)
```
same for testapi.Apps.ResourcePath

liyinan926 replied:
Ditto. We cannot use `testapi.Apps` as it defaults to `apps/v1beta1`.

liggitt commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -1738,3 +1740,17 @@ func TestRemoveCondition(t *testing.T) {
 }
+
+func resourcePath(resource, namespace, name string) string {
+	path := fmt.Sprintf("/apis/%s/%s/", apps.SchemeGroupVersion.Group, apps.SchemeGroupVersion.Version)
```
resourcePath being hardcoded to the apps group is unintuitive... I'd expect the testapi resource path helper to be used. if there are issues with that helper, can we resolve those issues instead?

liggitt replied:
> @liggitt do you have concerns about this?

not particularly, now that GKE has apps/v1beta2 enabled. this helper function should be named to make it clear that it is only dealing with apps/v1beta2

janetkuo replied:
We're going to punt this to 1.9, since it'll cause problems in updates to HA clusters. See: https://github.com/kubernetes/kubernetes/issues/51049#issuecomment-324784202

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -183,7 +183,7 @@ func ClusterRoles() []rbac.ClusterRole {
-				rbac.NewRule(ReadWrite...).Groups(extensionsGroup).Resources("daemonsets",
+				rbac.NewRule(ReadWrite...).Groups(extensionsGroup, appsGroup).Resources("daemonsets",
 					"deployments", "deployments/scale", "deployments/rollback", "ingresses",
 					"replicasets", "replicasets/scale", "replicationcontrollers/scale").RuleOrDie(),
```
replicationcontrollers/scale is not in the appsGroup... I would add the replicasets and replicasets/scale resources to the appsGroup rule on line 179 instead, alongside deployments

liyinan926 replied:
Done.

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -215,7 +215,7 @@ func ClusterRoles() []rbac.ClusterRole {
 				rbac.NewRule(ReadWrite...).Groups(batchGroup).Resources("jobs", "cronjobs").RuleOrDie(),
 
-				rbac.NewRule(ReadWrite...).Groups(extensionsGroup).Resources("daemonsets",
+				rbac.NewRule(ReadWrite...).Groups(extensionsGroup, appsGroup).Resources("daemonsets",
```
same here, this list includes things that do not exist in the apps group

liyinan926 replied:
Done.

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -239,7 +239,7 @@ func ClusterRoles() []rbac.ClusterRole {
 				rbac.NewRule(Read...).Groups(batchGroup).Resources("jobs", "cronjobs").RuleOrDie(),
 
-				rbac.NewRule(Read...).Groups(extensionsGroup).Resources("daemonsets", "deployments", "deployments/scale",
+				rbac.NewRule(Read...).Groups(extensionsGroup, appsGroup).Resources("daemonsets", "deployments", "deployments/scale",
```
and here

liyinan926 replied:
Done.

liggitt commented on test/e2e/apps/deployment.go:
```diff
@@ -25,6 +25,7 @@ import (
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
 
+	apps "k8s.io/api/apps/v1beta2"
```
do these e2e tests run against old clusters? if so, do we need to gate this test to only run on 1.8 and newer clusters?

liggitt replied:
hmm, nevermind... I see this is sticking with extensions/v1beta1 replicasets... when we switch this test to apps/v1beta2 ones, we might have to gate

liggitt commented on test/e2e/apps/replica_set.go:
```diff
@@ -35,7 +36,37 @@ import (
 )
 
-func newRS(rsName string, replicas int32, rsPodLabels map[string]string, imageName string, image string) *extensions.ReplicaSet {
+func newRS(rsName string, replicas int32, rsPodLabels map[string]string, imageName string, image string) *apps.ReplicaSet {
```
ok, here's where we may need to gate now... tests using apps.ReplicaSet won't run on old clusters, right?

liyinan926 replied:
I'm not sure. Just being curious, why do we have to gate? Is it because `apps/v1beta2` may not be enabled in old e2e clusters?

liggitt replied:
a 1.7 cluster will definitely not have apps/v1beta2 replicasets, right? @kubernetes/sig-testing-pr-reviews, am I correct that version skew tests can run e2e tests from master against prior cluster versions? I'm assuming the answer is yes based on the presence of things like https://github.com/kubernetes/kubernetes/blob/1957e4a2ced19d0beb85da9c232b8eb7ce2e771b/test/e2e/auth/service_accounts.go#L246

liyinan926 replied:
@kow3ns mentioned the e2e test should used the untyped client so it should theoretically work on both old and new e2e clusters.

liyinan926 replied:
Just chatted with @kow3ns and he had a good point: we should keep the e2e tests unchanged and if this PR is doing the right changes, the e2e tests using `extensions/vebeta1` should still work. The e2e tests can be updated when we are ready to update kubectl.

liggitt commented:
>  Is it better to have changes to pkg/kubectl in a separate PR

a 1.8 kubectl must continue working against a 1.7 API server, so it cannot exclusively switch to using apps/v1beta2

liyinan926 commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -1738,3 +1740,17 @@ func TestRemoveCondition(t *testing.T) {
 }
+
+func resourcePath(resource, namespace, name string) string {
+	path := fmt.Sprintf("/apis/%s/%s/", apps.SchemeGroupVersion.Group, apps.SchemeGroupVersion.Version)
```
IMO, without a big refactoring of testapi.go, we cannot use it.

liggitt replied:
as is, this test breaks in environments where v1beta2 is not enabled (like in GKE prior to today)

janetkuo replied:
Since this PR tries to switch RS controller to use apps/v1beta2 APIs, the controller itself wouldn't work without apps/v1beta2 being enabled

liyinan926 committed:
Move ReplicaSet controller to use apps/v1beta2 APIs

liyinan926 commented:
/retest

liyinan926 commented:
/retest

liyinan926 commented:
@liggitt PR updated. I addressed the comments and also reverted most changes to the e2e tests so they are still using `extensions` APIs. PTAL. Thanks!

janetkuo commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -42,22 +42,24 @@ import (
 	"k8s.io/kubernetes/pkg/securitycontext"
 )
 
+var codec = scheme.Codecs.LegacyCodec(apps.SchemeGroupVersion)
```
Name it `appsv1beta2codec`?

liyinan926 replied:
Done.

dixudx commented:
/test pull-kubernetes-unit

liyinan926 commented:
/keep-open

dims commented:
@liyinan926 are we trying to get this into 1.9? please rebase if so
